File: web/src/types/discover.ts
export type CatalogFilterOption = { value: string; label: string }

export type CatalogProvider = {
  id: string
  identity_namespaces: string[]
  operations: string[]
  media_types: string[]
  filters: string[]
  sorts: string[]
  operation_filters?: Record<string, string[]>
  operation_sorts?: Record<string, string[]>
  filter_options: Record<string, CatalogFilterOption[]>
}

export type MediaSubjectSummary = {
  media_subject_id: string
  media_type: 'movie' | 'series'
  canonical_title: string
  release_year: number | null
  poster_url: string | null
  provider_id: string
  external_id: string
  external_ids: Record<string, string>
  followed: boolean
  watchlisted: boolean
  degraded: boolean
}

export type DiscoverPageResponse = {
  items: MediaSubjectSummary[]
  continuation_token: string | null
  provider_id: string
  degraded: boolean
  cached_at: string | null
}

export type MediaSubjectDetail = MediaSubjectSummary & {
  original_title: string | null
  overview: string | null
  release_date: string | null
  genres: string[]
  regions: string[]
  rating: number | null
  rating_provider: string | null
  vote_count: number | null
  backdrop_url: string | null
  image_urls: string[]
  snapshot_updated_at: string | null
}

export type WatchlistPageResponse = {
  items: MediaSubjectSummary[]
  count: number
}

File: sundarr/app/schemas/discover.py
"""媒体发现 Core 的公共 API Schema。"""

from datetime import date, datetime

from pydantic import BaseModel, Field


class CatalogFilterOptionResponse(BaseModel):
    value: str
    label: str


class CatalogAttributionResponse(BaseModel):
    provider_name: str
    homepage_url: str
    notice: str
    logo_url: str | None = None


class CatalogProviderResponse(BaseModel):
    id: str
    identity_namespaces: list[str]
    operations: list[str]
    media_types: list[str]
    filters: list[str]
    sorts: list[str]
    operation_filters: dict[str, list[str]] = Field(default_factory=dict)
    operation_sorts: dict[str, list[str]] = Field(default_factory=dict)
    attribution: CatalogAttributionResponse | None = None
    filter_options: dict[str, list[CatalogFilterOptionResponse]] = Field(default_factory=dict)


class MediaSubjectSummary(BaseModel):
    media_subject_id: str
    media_type: str
    canonical_title: str
    release_year: int | None = None
    poster_url: str | None = None
    provider_id: str
    external_id: str
    external_ids: dict[str, str] = Field(default_factory=dict)
    followed: bool = False
    watchlisted: bool = False
    degraded: bool = False


class DiscoverPageResponse(BaseModel):
    items: list[MediaSubjectSummary]
    continuation_token: str | None = None
    provider_id: str
    degraded: bool = False
    cached_at: datetime | None = None


class MediaSubjectDetail(MediaSubjectSummary):
    original_title: str | None = None
    overview: str | None = None
    release_date: date | None = None
    genres: list[str] = Field(default_factory=list)
    regions: list[str] = Field(default_factory=list)
    rating: float | None = None
    rating_provider: str | None = None
    vote_count: int | None = None
    backdrop_url: str | None = None
    image_urls: list[str] = Field(default_factory=list)
    snapshot_updated_at: datetime | None = None


class FollowResponse(BaseModel):
    media_subject_id: str
    followed: bool
    followed_at: datetime | None = None


class WatchlistSyncResponse(BaseModel):
    provider_id: str
    pulled_count: int
    next_cursor: str | None = None
    last_synced_at: datetime


class WatchlistPageResponse(BaseModel):
    items: list[MediaSubjectSummary]
    count: int
